File: 1306-jump-game-iii/1306-jump-game-iii.java
class Solution {

    int n;
    int visited[];

    boolean solution(int i,int arr[]){
        if(i<0 || i>=n) return false;
        if(arr[i]==0) return true;
        if(visited[i]==1) return false;
        visited[i]=1;
        return solution(i-arr[i],arr)||solution(i+arr[i],arr);
    }

    public boolean canReach(int[] arr, int start) {
        n=arr.length;
        visited=new int[n];
        Arrays.fill(visited,0);
        return solution(start,arr);
    }
}

File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:

    int n;
    vector<int>visited;

    bool solution(int i,vector<int>&arr){
        if(i<0 || i>=n) return false;
        if(arr[i]==0) return true;
        if(visited[i]) return false;

        visited[i]=1;
        bool left=false,right=false;
        if(i+arr[i]<n) right=solution(i+arr[i],arr);
        if(i-arr[i]>=0) left=solution(i-arr[i],arr);

        return left||right;
    }

    bool canReach(vector<int>& arr, int start) {
        n=arr.size();
        visited.assign(n,0);
        return solution(start,arr);
    }
};